### ISSUE TITLE
UnexpectedValueException: Log file not available

### ISSUE THREAD

**Issue by Endeeman** [nextcloud/server#13285]
### Steps to reproduce
1. Log in as admin
2. Choose settings -> logging (from administration menu)
3. Select the 3 dots next to "Level" and click on "Download logs"

### Expected behaviour
Server should collect logging data and offer result as a downloadable file

### Actual behaviour
Error comes up - Internal sever error (on screen - see attachment). Logging says:
----------------------------
[index] Error: UnexpectedValueException: Log file not available at <<closure>>

0. /var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php line 166
   download()
1. /var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php line 99
   executeController(OC\Settings\Cont ... {}, "download")
2. /var/www/nextcloud/lib/private/AppFramework/App.php line 118
   dispatch(OC\Settings\Cont ... {}, "download")
3. /var/www/nextcloud/lib/private/AppFramework/Routing/RouteActionHandler.php line 47
   main("OC\\Settings\\C ... r", "download", OC\AppFramework\ ... {}, {_route: "settings.LogSettings.download"})
4. <<closure>>
   __invoke({_route: "settings.LogSettings.download"})
5. /var/www/nextcloud/lib/private/Route/Router.php line 297
   call_user_func(OC\AppFramework\ ... {}, {_route: "settings.LogSettings.download"})
6. /var/www/nextcloud/lib/base.php line 987
   match("/settings/admin/log/download")
7. /var/www/nextcloud/index.php line 42
   handleRequest()

GET /index.php/settings/admin/log/download
from 10.8.0.6 by XXXXXX at 2018-12-29T09:27:09+00:00
-----------------------------------------------------------

![nextcloud_fehler_download_protokoll](https://user-images.githubusercontent.com/46241099/50536776-586bd400-0b58-11e9-870a-35cab8b9deb5.png)

### Server configuration

**Operating system**: Ubuntu Server 18.04 LTS

**Web server:** Apache 2.4

**Database:** MariaDB

**PHP version:** 7

**Nextcloud version:** 15

**Updated from an older Nextcloud/ownCloud or fresh install:** Fresh Install

**Where did you install Nextcloud from:** Via wget -c https://download.nextcloud.com/server/releases/nextcloud-15.0.0.tar.bz2

**Signing status:**
<details>
<summary>Signing status</summary>

```
Login as admin user into your Nextcloud and access 
http://example.com/index.php/settings/integrity/failed 
paste the results here.

"No errors have been found."
```
</details>

**List of activated apps:**
Enabled:
  - accessibility: 1.1.0
  - activity: 2.8.2
  - bruteforcesettings: 1.3.0
  - calendar: 1.6.4
  - cloud_federation_api: 0.1.0
  - comments: 1.5.0
  - contacts: 2.1.8
  - dav: 1.8.0
  - federatedfilesharing: 1.5.0
  - federation: 1.5.0
  - files: 1.10.0
  - files_pdfviewer: 1.4.0
  - files_sharing: 1.7.0
  - files_texteditor: 2.7.0
  - files_trashbin: 1.5.0
  - files_versions: 1.8.0
  - files_videoplayer: 1.4.0
  - firstrunwizard: 2.4.0
  - gallery: 18.2.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.3.0
  - nextcloud_announcements: 1.4.0
  - notifications: 2.3.0
  - oauth2: 1.3.0
  - password_policy: 1.5.0
  - provisioning_api: 1.5.0
  - serverinfo: 1.5.0
  - sharebymail: 1.5.0
  - support: 1.0.0
  - survey_client: 1.3.0
  - systemtags: 1.5.0
  - tasks: 0.9.8
  - theming: 1.6.0
  - twofactor_backupcodes: 1.4.1
  - updatenotification: 1.5.0
  - workflowengine: 1.5.0
Disabled:
  - admin_audit
  - encryption
  - files_external
  - mail
  - user_external
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "XXX.XXX.XXX.XXX",
            "10.8.0.1"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "15.0.0.10",
        "overwrite.cli.url": "https:\/\/XXX.XXX.XXX.XXX",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "memcache.distributed": "\\OC\\Memcache\\Memcached",
        "memcached_servers": [
            [
                "localhost",
                11211
            ]
        ],
        "maintenance": false,
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpauth": 1,
        "mail_smtpsecure": "tls",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***"
    }
}

```
</details>

**Are you using encryption:** not inside Nextcloud but VPN is used to access Website
</details>

### Client configuration
**Browser:** Firefox 64

**Operating system:** Ubuntu 18.04 LTS

### Logs
#### Web server error log
<details>
<summary>Web server error log</summary>
[Sat Dec 29 06:25:02.816606 2018] [mpm_prefork:notice] [pid 1457] AH00163: Apache/2.4.29 (Ubuntu) OpenSSL/1.1.0g configured -- resuming normal operations
[Sat Dec 29 06:25:02.816655 2018] [core:notice] [pid 1457] AH00094: Command line: '/usr/sbin/apache2'
[Sat Dec 29 09:26:18.746902 2018] [authz_core:error] [pid 4038] [client 10.8.0.6:58312] AH01630: client denied by server configuration: /var/www/nextcloud/data/.ocdata
[Sat Dec 29 09:26:23.686701 2018] [authz_core:error] [pid 4038] [client 10.8.0.6:58312] AH01630: client denied by server configuration: /var/www/nextcloud/data/.ocdata
[Sat Dec 29 09:29:54.653250 2018] [authz_core:error] [pid 5327] [client 10.8.0.6:58338] AH01630: client denied by server configuration: /var/www/nextcloud/data/.ocdata
[Sat Dec 29 09:33:19.011158 2018] [negotiation:error] [pid 5328] [client 10.8.0.6:58388] AH00687: Negotiation: discovered file(s) matching request: /var/www/nextcloud/settings/help (None could be negotiated).
[Sat Dec 29 09:37:37.019000 2018] [authz_core:error] [pid 5327] [client 10.8.0.6:58468] AH01630: client denied by server configuration: /var/www/nextcloud/data/.ocdata
[Sat Dec 29 09:37:53.993572 2018] [authz_core:error] [pid 5527] [client 10.8.0.6:58480] AH01630: client denied by server configuration: /var/www/nextcloud/data/.ocdata
[Sat Dec 29 09:38:00.182929 2018] [authz_core:error] [pid 5527] [client 10.8.0.6:58480] AH01630: client denied by server configuration: /var/www/nextcloud/data/.ocdata
[Sat Dec 29 09:49:43.746470 2018] [authz_core:error] [pid 4431] [client 10.8.0.6:58766] AH01630: client denied by server configuration: /var/www/nextcloud/data/.ocdata
[Sat Dec 29 10:02:00.702756 2018] [authz_core:error] [pid 5346] [client 10.8.0.6:58864] AH01630: client denied by server configuration: /var/www/nextcloud/data/.ocdata
[Sat Dec 29 10:06:48.674165 2018] [authz_core:error] [pid 4365] [client 10.8.0.6:58904] AH01630: client denied by server configuration: /var/www/nextcloud/data/.ocdata
[Sat Dec 29 10:20:27.566452 2018] [authz_core:error] [pid 4430] [client 10.8.0.6:58982] AH01630: client denied by server configuration: /var/www/nextcloud/data/.ocdata
[Sat Dec 29 10:20:31.506168 2018] [authz_core:error] [pid 4430] [client 10.8.0.6:58982] AH01630: client denied by server configuration: /var/www/nextcloud/data/.ocdata
[Sat Dec 29 10:22:01.136419 2018] [authz_core:error] [pid 5753] [client 10.8.0.6:59010] AH01630: client denied by server configuration: /var/www/nextcloud/data/.ocdata
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>
{"reqId":"W7ZryEV4kP4zjubEfkI4","level":3,"time":"2018-12-29T10:21:49+00:00","remoteAddr":"10.8.0.6","user":"XXXXXXXX","app":"index","method":"GET","url":"\/index.php\/settings\/admin\/log\/download","message":{"Exception":"UnexpectedValueException","Message":"Log file not available","Code":0,"Trace":[{"file":"\/var\/www\/nextcloud\/lib\/private\/AppFramework\/Http\/Dispatcher.php","line":166,"function":"download","class":"OC\\Settings\\Controller\\LogSettingsController","type":"->","args":[]},{"file":"\/var\/www\/nextcloud\/lib\/private\/AppFramework\/Http\/Dispatcher.php","line":99,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OC\\Settings\\Controller\\LogSettingsController"},"download"]},{"file":"\/var\/www\/nextcloud\/lib\/private\/AppFramework\/App.php","line":118,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OC\\Settings\\Controller\\LogSettingsController"},"download"]},{"file":"\/var\/www\/nextcloud\/lib\/private\/AppFramework\/Routing\/RouteActionHandler.php","line":47,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OC\\Settings\\Controller\\LogSettingsController","download",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"_route":"settings.LogSettings.download"}]},{"function":"__invoke","class":"OC\\AppFramework\\Routing\\RouteActionHandler","type":"->","args":[{"_route":"settings.LogSettings.download"}]},{"file":"\/var\/www\/nextcloud\/lib\/private\/Route\/Router.php","line":297,"function":"call_user_func","args":[{"__class__":"OC\\AppFramework\\Routing\\RouteActionHandler"},{"_route":"settings.LogSettings.download"}]},{"file":"\/var\/www\/nextcloud\/lib\/base.php","line":987,"function":"match","class":"OC\\Route\\Router","type":"->","args":["\/settings\/admin\/log\/download"]},{"file":"\/var\/www\/nextcloud\/index.php","line":42,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"\/var\/www\/nextcloud\/settings\/Controller\/LogSettingsController.php","Line":59,"CustomMessage":"--"},"userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:64.0) Gecko\/20100101 Firefox\/64.0","version":"15.0.0.10"}
</details>

**Comment by mr-bolle** [comment]
```
$ ls -la /var/www/nextcloud/data/.ocdata  
-rwxr-xr-x 1 www-data www-data 0 Dec 20 13:53 data/.ocdata
````
if this file is owned by the user root, you can correct this

`chown www-data:www-data /var/www/nextcloud/data/.ocdata`

**Comment by andyxh** [comment]
I'm getting same error

**Comment by Endeeman** [comment]
> ```
> $ ls -la /var/www/nextcloud/data/.ocdata  
> -rwxr-xr-x 1 www-data www-data 0 Dec 20 13:53 data/.ocdata
> ```
> 
> if this file is owned by the user root, you can correct this
> 
> `chown www-data:www-data /var/www/nextcloud/data/.ocdata`

@mr-bolle: File is already owned by www-data. So chown will have no effect, am I right?
-rw-r--r--  1 www-data www-data        0 Dez 14 08:39 .ocdata

I'm not sure whether .ocdata shouldn't contain any data...?!

**Comment by andyxh** [comment]
so you have no more issues ?

**Comment by Endeeman** [comment]
no more issues...

**Comment by andyxh** [comment]
I'm still having issues

**Comment by Endeeman** [comment]
still investigating...
Did someone saw this: https://help.nextcloud.com/t/clean-nc15-install-cant-download-logs-internal-server-error/44028/6

**Comment by danielkesselberg** [comment]
@rullzer introduced by https://github.com/nextcloud/server/commit/9319d557a4b1156d82d824cdb1b7d6f07ad3fbde not sure how to fix this. Before a instance of `OC\Log` has been passed to `LogSettingsController` after `OC\AppFramework\Logger`.

**Comment by danielkesselberg** [comment]
```
	public function __construct(string $appName, IRequest $request, IServerContainer $c) {
		parent::__construct($appName, $request);
		$this->log = $c->getLogger();
	}
```

With the above change for `LogSettingsController` it works for me again but not sure if the way `IServerContainer` is injected here is correct.

**Comment by danielkesselberg** [comment]
```
Index: settings/Controller/LogSettingsController.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- settings/Controller/LogSettingsController.php	(revision aa00b2b277e9b8d831b5381b9dc5a2cd92212182)
+++ settings/Controller/LogSettingsController.php	(date 1546116198000)
@@ -31,6 +31,7 @@
 use OCP\AppFramework\Http\StreamResponse;
 use OCP\ILogger;
 use OCP\IRequest;
+use OCP\IServerContainer;
 
 /**
  * Class LogSettingsController
@@ -42,9 +43,9 @@
 	/** @var ILogger */
 	private $log;
 
-	public function __construct(string $appName, IRequest $request, ILogger $logger) {
+	public function __construct(string $appName, IRequest $request, IServerContainer $c) {
 		parent::__construct($appName, $request);
-		$this->log = $logger;
+		$this->log = $c->getLogger();
 	}
 
 	/**
```

You may add `use OCP\IServerContainer;` below `use OCP\IRequest;`

**Comment by Endeeman** [comment]
Oh no - getting tired :-(
You're completely right!
"use OCP\IServerContainer;" was missing...
Thx a lot!

**Comment by danielkesselberg** [comment]
We should leave this ticket open until fixed by a pull request.

**Comment by rullzer** [comment]
Ah my bad indeed. Let me tackle this soonish

**Comment by rullzer** [comment]
Fix in #13402
